dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/aggregation/PipelineResultIteration.java

```diff
@@ -0,0 +1,132 @@
+     *  result documents to equivalent variable names in the original query.
+     *  Where an entry does not exist for a field, the field name and variable
+     *  name are assumed to be the same.
+     * @param bindings A partial solution. May be empty.
+     */
+    public PipelineResultIteration(AggregateIterable<Document> aggIter,
+            Map<String, String> varToOriginalName,
+            BindingSet bindings) {
+        aggIter.batchSize(BATCH_SIZE);
+        this.cursor = aggIter.iterator();
+        this.varToOriginalName = varToOriginalName;
+        this.bindings = bindings;
+        lookahead();
+    }
+
+    private void lookahead() {
+        while (nextSolution == null && cursor.hasNext()) {
+            nextSolution = docToBindingSet(cursor.next());
+        }
+    }
+
+    @Override
+    public boolean hasNext() throws QueryEvaluationException {
+        lookahead();
+        return nextSolution != null;
+    }
+
+    @Override
+    public BindingSet next() throws QueryEvaluationException {
+        lookahead();
+        BindingSet solution = nextSolution;
+        nextSolution = null;
+        return solution;
+    }
+
+    @Override
+    public void remove() throws QueryEvaluationException {
+        lookahead();
+        nextSolution = null;
+    }
+
+    @Override
+    public void close() throws QueryEvaluationException {
+        cursor.close();
+    }
+
+    private QueryBindingSet docToBindingSet(Document result) {
+        QueryBindingSet bindingSet = new QueryBindingSet(bindings);
+        Document valueSet = result.get(AggregationPipelineQueryNode.VALUES, Document.class);
+        Document typeSet = result.get(AggregationPipelineQueryNode.TYPES, Document.class);
+        if (valueSet != null) {
+            for (Map.Entry<String, Object> entry : valueSet.entrySet()) {
+                String fieldName = entry.getKey();
+                String valueString = entry.getValue().toString();
+                String typeString = typeSet == null ? null : typeSet.getString(fieldName);
+                String varName = varToOriginalName.getOrDefault(fieldName, fieldName);
+                Value varValue;
+                if (typeString == null || typeString.equals(XMLSchema.ANYURI.stringValue())) {
+                    varValue = VF.createURI(valueString);
+                }
+                else {
+                    varValue = VF.createLiteral(valueString, VF.createURI(typeString));
+                }
+                Binding existingBinding = bindingSet.getBinding(varName);
+                // If this variable is not already bound, add it.
+                if (existingBinding == null) {
+                    bindingSet.addBinding(varName, varValue);
+                }
+                // If it's bound to something else, the solutions are incompatible.
+                else if (!existingBinding.getValue().equals(varValue)) {
+                    return null;
```

Comment:
Should this throw an exception? Why does it return null?

Reply:
This would cause your Iteration to periodically return null which will likely result in a NPE by whoever is using it since that would be pretty unexpected.

Reply:
This method is only called internally, by lookahead(). It returns null when the next solution returned by the pipeline is incompatible with any bindings that were passed into the constructor. (The constructor may have received a non-empty BindingSet if the pipeline query is being evaluated in the context of a specific partial solution.) The lookahead method will use this to loop through pipeline results until it finds one that is compatible with the bindings, or until the pipeline yields no more results. The null value should never make its way to the caller. I didn't use an exception here because it isn't really an exceptional case in that context: we should expect some of the solutions to the pipeline subquery to not match the specific candidate solution.

Skipping over non-matching results client-side might be inefficient, and we could implement this instead by adding a "$match" step for the candidate solution onto the pipeline. But this wouldn't solve the main inefficiency which is that either way we're executing the pipeline subquery for each candidate solution. So the approach I've taken here is just to do something that satisfies the API correctly, and hope that we can structure queries to avoid the inherent inefficiency. (For example, if we had `Join(Join(SP1, SP2), Join(SP3, SP4))`, the visitor as currently implemented would turn that into a join of two pipeline nodes and have to pass binding sets from one into the other, which will be slow regardless of how that's done. But if it's instead parsed as `Join(Join(Join(SP1, SP2), SP3), SP4))` then we can translate the whole thing into one pipeline query, which should be faster.)

I added a test for this iterator which includes the case of a provided BindingSet and incompatible partial solutions, verifying that the expected (non-null) results are returned.

Reply:
Oh I see how that's being used now. Thanks for the explanation.